sdk/identity/azure-identity/src/token_credential_impl.cpp

```diff
@@ -161,140 +152,52 @@ AccessToken TokenCredentialImpl::ParseToken(
+    accessToken.Token = parsedJson[accessTokenPropertyName].get<std::string>();
 
-  int64_t value = 0;
-  {
-    auto const size = jsonString.size();
-    for (auto pos = valueStartPos; pos < size; ++pos)
+    if (parsedJson.contains(expiresInPropertyName))
     {
-      auto c = jsonString[pos];
-      if (c < '0' || c > '9')
-      {
-        break;
-      }
-
-      value = (value * 10) + (static_cast<int64_t>(c) - '0');
+      accessToken.ExpiresOn
+          = std::chrono::system_clock::now()
+          + std::chrono::seconds(
+                parsedJson[expiresInPropertyName].get<std::chrono::seconds::duration::rep>());
     }
-  }
-
-  outValue = value;
-
-  return true;
-}
-
-std::string::size_type GetPropertyValueEnd(std::string const& str, std::string::size_type startPos);
-
-bool GetPropertyValueAsString(
-    std::string const& jsonString,
-    std::string const& propertyName,
-    std::string& outValue)
-{
-  auto const valueStartPos = GetPropertyValueStart(jsonString, propertyName);
-  if (valueStartPos == std::string::npos)
-  {
-    return false;
-  }
-  auto const jsonStringBegin = jsonString.begin();
-  outValue = std::string(
-      jsonStringBegin + valueStartPos,
-      jsonStringBegin + GetPropertyValueEnd(jsonString, valueStartPos));
-
-  return true;
-}
-
-std::string::size_type GetPropertyValueStart(
-    std::string const& jsonString,
-    std::string const& propertyName)
-{
-  auto const propertyNameValueSeparator = jsonString.find(':', jsonString.find(propertyName));
-  if (propertyNameValueSeparator == std::string::npos)
-  {
-    return std::string::npos;
-  }
-
-  auto pos = propertyNameValueSeparator;
-  {
-    auto const jsonStringSize = jsonString.size();
-    for (; pos < jsonStringSize; ++pos)
+    else if (expiresOnPropertyName.empty())
     {
-      auto c = jsonString[pos];
-      if (c != ':' && c != ' ' && c != '\"' && c != '\'')
-      {
-        break;
-      }
+      ThrowMissingJsonPropertyError(expiresInPropertyName);
     }
-  }
-
-  return pos;
-}
-
-std::string::size_type GetPropertyValueEnd(std::string const& str, std::string::size_type startPos)
-{
-  auto pos = startPos;
-  {
-    auto const strSize = str.size();
-    for (; pos < strSize; ++pos)
+    else if (!parsedJson.contains(expiresOnPropertyName))
+    {
+      ThrowMissingJsonPropertyError(expiresOnPropertyName);
+    }
+    else
     {
-      auto c = str[pos];
-      if (c == '\"' || c == '\'')
+      auto expiresOn = parsedJson[expiresOnPropertyName].get<std::string>();
       {
-        break;
+        auto const spacePos = expiresOn.find(' ');
+        if (spacePos != std::string::npos) 
+        {
+          expiresOn = expiresOn.replace(spacePos, 1, 1, 'T');
+        }
       }
+
+      accessToken.ExpiresOn
+          = Azure::DateTime::Parse(expiresOn, Azure::DateTime::DateFormat::Rfc3339);
     }
-  }
 
-  return pos;
+    return accessToken;
+  }
+  catch (Azure::Core::Json::_internal::json::parse_error const& ex)
```

Comment:
Is the whole purpose of this try-catch to avoid leaking the internal nlohmann JSON's parse exception, or is there some other reason?

Is this a pattern we want to follow everywhere we do JSON parsing? I think we have places where the error just bubbles up, without us wrapping the code in a try-catch and our own exception.

Reply:
Since we use nlohmann json internally, I think we should be wrapping these exceptions.
Speaking of authentication space, that is even more important, because if something else other than `AuthenticationException` is getting thrown, it becomes hard to distinguish whether the exception came from the request, or from the authentication pipeline.